glommio/src/parking.rs

```diff
@@ -692,14 +763,51 @@ impl Reactor {
 // be allowed in files that don't support it, and same for readable() writable()
 impl Source {
     pub(crate) async fn collect_rw(&self) -> io::Result<usize> {
-        future::poll_fn(|cx| {
-            if let Some(result) = self.take_result() {
-                return Poll::Ready(result);
+        self.collect_io_stats(
+            future::poll_fn(|cx| {
+                if let Some(result) = self.take_result() {
+                    return Poll::Ready(result);
+                }
+
+                self.add_waiter(cx.waker().clone());
+                Poll::Pending
+            })
+            .await,
+        )
+    }
+
+    fn collect_io_stats(&self, io_result: io::Result<usize>) -> io::Result<usize> {
```

Comment:
I am not sure this is the right place to collect these.

For buffered file streams, for instance, we not always call `collect_rw`.  (see the `poll_xxx` functions for BufferedFile)

What bothers me the most, though, is that now we have a jump table at the end of a potentially very fast I/O operation. That's God's way of telling you the abstraction level is wrong! At the right abstraction layer we know which operation type we are accounting so we don't need to jump

Sockets will be a bit trickier because there is a `Source` path and a `yolo` path, but we're also interested in knowing how many times we're succeeding in the yolo path, so that can be accounted separately (plus sockets are not covered in this PR)

Reply:
Thanks for the feedback! I've reworked the patch and changed my approach such that:
- We no longer have a jump table on the hot path (we know exactly what counter to increment)
- We record stats for each individual rings (we also provide a convenient way to get the stats for all rings)

Reply:
Note that this work doesn't tackle sockets at all but we can add them later. The way things are set up, it should be easy (even for the yolo path that issue syscalls directly).